### PR TITLE
Segwit serialization support and nested p2sh signing

### PR DESCRIPTION
**Pull request by braydonf**
Adds initial segwit support, including:
- New transaction serialization that includes witness data
- Nested p2sh segwit transaction creation, signing and verification _(Ability to use segwit with p2sh multisig addresses.)_
- Verification of segwit transaction scripts
  - pay-to-witness-public-key-hash
  - pay-to-witness-script-hash
  - pay-to-nested-p2sh-witness

Closes: https://github.com/bitpay/bitcore-lib/issues/49
Closes: https://github.com/bitpay/bitcore-lib/issues/42

### CODE REVIEW

File: lib/transaction/input/multisigscripthash.js [hunk 1]
```diff
@@ -62,6 +77,38 @@ MultiSigScriptHashInput.prototype._serializeSignatures = function() {
...
+    var redeemScriptBuffer = this.redeemScript.toBuffer();
+    writer.writeVarintNum(redeemScriptBuffer.length);
+    writer.write(redeemScriptBuffer);
+  } else {
+    throw new Error('@TODO');
+  }
+  return writer.toBuffer();
+};
+
+MultiSigScriptHashInput.prototype.getSatoshisBuffer = function() {
```
Comment (afk11) [inline lib/transaction/input/multisigscripthash.js]: Maybe move this to input? It helps keep pay-to-script-hash/witness-scripthash handling more general

**Comment by bityogi** [comment]
+1. I'll be following this one closely. Thanks for working on this.

**Comment by rubensayshi** [comment]
@braydonf what's the status on this PR and if it's not finished what is left to do and how can I help?

**Comment by braydonf** [comment]
@rubensayshi from what I remember, I think these need to be finished:

- Adding coverage by pulling test cases from: https://github.com/bitcoin/bitcoin/blob/master/src/test/transaction_tests.cpp#L422-L612 *(nested witness p2sh is only tested currently)*. There could be additional changes necessary to get the tests to pass.
- Cleanup and refactoring
- Remove segnet network and test with testnet, and update for any changes that have happened since segnet3
- Code and API review. For example witness serialization should likely be enabled if the transaction is segwit for example by default. I think it requires specifying the serialization currently.

File: lib/address.js [hunk 2]
```diff
@@ -255,11 +255,16 @@ Address._transformScript = function(script, network) {
  * @param {Array} publicKeys - a set of public keys to create an address
  * @param {number} threshold - the number of signatures needed to release the funds
  * @param {String|Network} network - either a Network instance, 'livenet', or 'testnet'
+ * @param {boolean=} nestedWitness - if the address uses a nested p2sh witness
  * @return {Address}
  */
-Address.createMultisig = function(publicKeys, threshold, network) {
+Address.createMultisig = function(publicKeys, threshold, network, nestedWitness) {
```
Comment (afk11) [inline lib/address.js]: Would Address.createP2SH be better here, allowing arbitrary scripts to be passed in? Technically, multisig outputs don't have an address, but P2SH outputs do

**Comment by afk11** [comment]
> Adding coverage by pulling test cases from: https://github.com/bitcoin/bitcoin/blob/master/src/test/transaction_tests.cpp#L422-L612 (nested witness p2sh is only tested currently). There could be additional changes necessary to get the tests to pass.

FWIW: https://github.com/bitcoin/bitcoin/blob/master/src/test/script_tests.cpp#L857 this might also be useful if you don't test against it already. It's mainly for verification purposes, but may turn up bugs outside of this PR. 

When working on signing, it's best to define your standard types (p2pk/p2pkh/multisig) and be able to deal with them anywhere P2SH, P2WSH comes in, since both of these can be used in tandem. An example of things maybe going awry is this file: https://github.com/bitpay/bitcore-lib/blob/master/lib/transaction/input/multisigscripthash.js & the multisig.js that was added after. P2SH & P2WSH require general modifications to a plain script-in-txout-script style transaction. AFAICT Bitcoin Core does this as well.

When signing any input, the redeemScript / witnessScripts / output amount are optional, sigVersion depends on witness version.. Then, observing that the contents of a scriptWitness is the same as a scriptStack after parsing the equivalent scriptSig..you can keep parsed/created signature's in an array until it's time to serialize, then decide where you put them (either as a Script, or a witness).

This also helps dealing with P2SH, which requires the redeemScript to always be present in your scriptSig (rather than having `p2shpubkeyhashinput.js` and `p2shpubkeyinput.js`). 

I would submit a patch, but it's not a trivial refactoring.. just my two cents worth!

File: lib/address.js (hunk 2 above)
Comment (braydonf) [inline lib/address.js]: `Address.payingTo` is for allowing arbitrary scripts to be passed in. `Address.createMultisig` is for creating multisig p2sh addresses, as bare multisig doesn't have an address, as you mention.

**Comment by braydonf** [comment]
@afk11 We have tests using https://github.com/bitpay/bitcore-lib/blob/master/test/data/bitcoind/script_invalid.json and https://github.com/bitpay/bitcore-lib/blob/master/test/data/bitcoind/script_valid.json however these should likely also be updated. Added to the list.

**Comment by btcdrak** [comment]
Note witness v0 program can be up to 40 bytes long https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#Witness_program

Only use compressed keys as per BIP143  https://github.com/bitcoin/bips/pull/459/files#diff-d6a595a37a5d316054e79e53e9c2382fR128

**Comment by afk11** [comment]
bump

**Comment by afk11** [comment]
What's the status of this PR?

**Comment by braydonf** [comment]
It's ready for testing, and can go on 0.13 or 1.0.0 branches.

**Comment by afk11** [comment]
I guess I mean, how are things going supporting [P2PK, P2PKH, multisig] in any representation of [P2SH,  P2WSH, P2SH|P2WSH]? Currently there's only a `buildWitnessMultisigOutFromScript`

**Comment by braydonf** [comment]
The scripts for pay-to-witness-public-key-hash and pay-to-witness-script-hash can be built and verified,  there are tests for this case. However I don't think there are convenience methods for them. It's related to the deferred BIP142 https://github.com/bitcoin/bips/blob/master/bip-0142.mediawiki

So I think adding native segwit support can be made with another PR, while this one can bridge in what is necessary to support the transaction serialization and script verification.

**Comment by afk11** [comment]
Ready to go?

**Comment by braydonf** [comment]
Rebased onto latest master. Previous branch is at https://github.com/braydonf/bitcore-lib/tree/segwit-2016-10-6 (to diff against).

**Comment by afk11** [comment]
What is the ETA for this to be merged? Does it currently sign all the previously supported script types with P2SH and P2SH|P2WSH?

**Comment by gabegattis** [comment]
@afk11 I would not expect to see any additional work on this PR until segwit activation appears imminent.

**Comment by SherlockStd** [comment]
Any news ?

**Comment by afk11** [comment]
Per an email to the Segwit 2x mailing list, support for this library is being discontinued (called it #119) in favor of bcoin.

It's genuinely frustrating I, and certainly @braydonf, put effort into adding segwit support for this library 19 months ago (with others offering to help), for it to be discontinued when people have been testing other implementations for as long. 

I guess that's what you get for using a business backed bitcoin implementation - and I will hesitate before I do so again (looking at you bcoin)

**Comment by lichtamberg** [comment]
Is there still progress or do I have to move away from copay/bitcore?

**Comment by afk11** [comment]
@lichtamberg looks like it. They're suggesting everyone upgrades to btc1 https://blog.bitpay.com/bitcore-segwit-activation/

**Comment by lichtamberg** [comment]
Lol, I would never use this shit BTC1.. And I'm shocked that bitpay is distributing malware now: https://twitter.com/larrybitcoin/status/898188496015425536 

Bitpay was a really nice company, but within the last months they went fully retard...

**Comment by braydonf** [comment]
FYI:  

Satoshi Labs, @runn1ng, has been maintaining a branch that supports segwit for quite some time now:
- https://github.com/satoshilabs/bitcoin

Also Bitcore patches for indexing were applied to 0.13 Bitcoin Core a year ago, which also supports segwit *(made sure that these were applied to support segwit when activated)*:
- https://github.com/bitpay/bitcoin/tree/0.13-bitcore

**Comment by afk11** [comment]
@braydonf 

Bitcore-node is really not the problem, bitcore-lib is. Bitcore-lib / this repository does not have the BIP143, BIP141 code, unless it's in this pull request (which is also incomplete).

**Comment by bobquest33** [comment]
Why this https://blog.bitpay.com/bitcore-segwit-activation/ was posted, what is the stand of bitpay and bitcore will it only support segwit2x. Has the orginal segwit implemented or not. Please clarify.

**Comment by runn1ng** [comment]
@afk11 if you care only about backend for insight/bitcore, this is enough, you don't need the code for creating segwit addresses.

Also you don't need that for copay either, it doesn't use segwit currently.

So if you don't plan to also update copay to use segwit addresses *for user's addresses* something like this is enough.

**Comment by braydonf** [comment]
An update on this:

There is a small downside to using a nested witness script; the transaction is actually _larger_ than direct P2WPKH or P2PKH, so it is less disk and bandwidth efficient. However it may still mean less fees, especially for the case of a large number of signers. It was unclear what the address would be for segwit at the time this was implemented.

Now with [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki) and [bech32](http://bitcoin.sipa.be/bech32/demo/demo.html), this is a better alternative to using a nested witness script. And there is a push to, at the minimum, support sending to bech32 addresses in more wallets: https://whensegwit.com/

**Pull request merged**